### ISSUE TITLE
YOLO Semantic Segmentation Predictions Produce Jagged Lines

### ISSUE THREAD

**Issue by maalik-marzouki1** [ultralytics/ultralytics#24823]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict, Train

### Bug

I am training a YOLO-based semantic segmentation model for tennis pitch line segmentation. The model successfully detects the pitch lines, but the predicted masks contain highly jagged and irregular boundaries instead of smooth, continuous lines.

###  Approach 1: Binary Segmentation (nc=1 Pitch Only)
png mask ( pixel value 0 =background , 1=pitch)
names:
  0: pitch

label_mapping:
  0: 0
  1: 1  

      Epoch    GPU_mem    ce_loss dice_pitch   aux_loss  Instances       Size
    154/500      7.08G    0.04084     0.2145    0.01617          4        640: 100% ━━━━━━━━━━━━ 307/307 2.6it/s 1:56<0.3ss
                 Class     Images     PixelsmIoU(pitch)PixAcc(all): 100% ━━━━━━━━━━━━ 33/33 3.8it/s 8.7s0.3s
                   all        522    3731714      0.566      0.987

EarlyStopping: Training stopped early as no improvement observed in last 100 epochs. Best results observed at epoch 54
To update EarlyStopping(patience=100) pass a new patience value, i.e. `patience=300` or use `patience=0` to disable EarlyStopping.

###  Approach 2: Binary Segmentation (nc=1 Pitch Only)
png mask ( pixel value 1 =background , 0=pitch)
names:
  0: pitch

label_mapping:
  0: 0
  1: 1  and I inverted the validation metric computation so that the reported mIoU corresponds only to the pitch class rather than the background.
  ### Training Results

At epoch 100:

```text
Epoch    GPU_mem    ce_loss dice_pitch   aux_loss  Instances       Size
100/500      7.08G    0.02149   0.006821     0.0159          4        640
```

Validation metrics:

```text
Class        Images      Pixels        mIoU       PixelAcc
all             522   213811202       0.538       0.988
pitch           522     3731714       0.538       0.819
background      522   210079488       0.988       0.991
```

###  Approach 3: Two-Class Segmentation (Pitch + Background) nc=2

Training configuration:

Class 0: pitch
Class 1: background

Because background represents approximately 95% of all pixels, I evaluated the model using a weighted mIoU metric:

Weighted mIoU = (10 × mIoU(pitch) + 1 × mIoU(background)) / 11

      Epoch    GPU_mem    ce_loss  dice_loss   aux_loss  Instances       Size
      7/500       7.2G    0.06995     0.1378    0.04657          4        640: 100% ━━━━━━━━━━━━ 307/307 2.0it/s 2:300.3ss
                 Class     Images     Pixels       mIoU     PixAcc: 100% ━━━━━━━━━━━━ 66/66 3.1it/s 21.3s0.3ss
                 Class     Images     Pixels       mIoU     PixAcc       Dice
                   all        522  213811202       0.55      0.983      0.673
                 pitch        522    3731714      0.507      0.976      0.673

<img width="640" height="640" alt="Image" src="https://github.com/user-attachments/assets/431d200d-2b38-4567-92a5-7eb4adcfda3b" />

### Environment

annotated-types==0.7.0
certifi==2026.4.22
charset-normalizer==3.4.7
click==8.3.3
colorama==0.4.6
contourpy==1.3.3
cuda-bindings==13.2.0
cuda-pathfinder==1.5.4
cuda-toolkit==13.0.2
cycler==0.12.1
filelock==3.29.0
flatbuffers==25.12.19
fonttools==4.62.1
fsspec==2026.4.0
gitdb==4.0.12
GitPython==3.1.50
idna==3.13
Jinja2==3.1.6
kiwisolver==1.5.0
MarkupSafe==3.0.3
matplotlib==3.10.9
ml_dtypes==0.5.4
mpmath==1.3.0
networkx==3.6.1
numpy==2.4.4
nvidia-cublas==13.1.0.3
nvidia-cuda-cupti==13.0.85
nvidia-cuda-nvrtc==13.0.88
nvidia-cuda-runtime==13.0.96
nvidia-cudnn-cu13==9.19.0.56
nvidia-cufft==12.0.0.61
nvidia-cufile==1.15.1.6
nvidia-curand==10.4.0.35
nvidia-cusolver==12.0.4.66
nvidia-cusparse==12.6.3.3
nvidia-cusparselt-cu13==0.8.0
nvidia-nccl-cu13==2.28.9
nvidia-nvjitlink==13.0.88
nvidia-nvshmem-cu13==3.4.5
nvidia-nvtx==13.0.85
onnx==1.21.0
onnxconverter-common==1.16.0
onnxruntime-gpu==1.26.0
onnxslim==0.1.94
opencv-python==4.13.0.92
packaging==26.2
pillow==12.2.0
platformdirs==4.9.6
polars==1.40.1
polars-runtime-32==1.40.1
protobuf==7.34.1
psutil==7.2.2
pydantic==2.13.4
pydantic_core==2.46.4
pyparsing==3.3.2
python-dateutil==2.9.0.post0
PyYAML==6.0.3
requests==2.33.1
scipy==1.17.1
sentry-sdk==2.59.0
setuptools==81.0.0
six==1.17.0
smmap==5.0.3
sympy==1.14.0
tensorrt_cu13==11.0.0.114
tensorrt_cu13_bindings==11.0.0.114
tensorrt_cu13_libs==11.0.0.114
torch==2.11.0
torchvision==0.26.0
tqdm==4.68.1
triton==3.6.0
typing-inspection==0.4.2
typing_extensions==4.15.0
-e git+ssh://git@github.com/ultralytics/ultralytics.git@ce5371c9d819b965f09c0985339b2964b48de336#egg=ultralytics
ultralytics-thop==2.0.19
urllib3==2.6.3
wandb==0.26.1

### Minimal Reproducible Example

names:
  0: pitch

label_mapping:
  0: 0
  1: 1

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @maalik-marzouki1, thank you for the detailed report and for your interest in Ultralytics 🚀! This is an automated response to help gather the key details quickly, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this appears to be a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it. In particular, it would help if you could share:
- The exact training command or script used
- The exact prediction command or script used
- A small sample image and its corresponding mask
- Whether the jagged mask boundaries appear in both validation outputs and standalone `predict` outputs
- Any custom changes made to loss, metrics, label mapping, or postprocessing

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

Is this using latest Ultralytics?

What's the output from Ultralytics prediction?

What's your training command?

**Comment by smqd19** [comment]
The jagged predictions could be due to the model not learning to smooth out the boundaries. Here are a few steps you can try:

- **Post-processing**: Apply a morphological operation like closing or opening to smooth out the jagged boundaries. This can be done using libraries like OpenCV or scikit-image.
- **Loss Function**: Ensure that you're using a loss function that penalizes jagged boundaries. For example, you can add a term to the loss function that penalizes the length of the boundary.
- **Data Augmentation**: Try adding more data augmentation techniques such as rotation, scaling, and flipping to help the model learn more robust features.
- **Model Architecture**: Consider using a more complex model architecture or adding additional layers to help the model learn smoother boundaries.

Here's a simple example of how you might apply a morphological operation using OpenCV:

```python
import cv2

# Load the predicted mask
mask = cv2.imread('predicted_mask.png', cv2.IMREAD_GRAYSCALE)

# Apply morphological closing
kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (5, 5))
smooth_mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel)

# Save the smoothed mask
cv2.imwrite('smoothed_mask.png', smooth_mask)
```

**Comment by glenn-jocher** [comment]
Thanks for the suggestion — for semantic segmentation this is usually a resolution/class-map limitation rather than missing smoothing: YOLO returns a dense `result.semantic_mask` and upsamples it back to the original image, so very thin court lines can look stair-stepped at `imgsz=640`; please retest with native `.pt` inference at `imgsz=1024` or `1280`, and if it still reproduces, share one image/mask pair plus the text output of `!yolo checks` so we can reproduce it locally. See [Semantic Segmentation](https://docs.ultralytics.com/tasks/semantic) and [`SemanticSegmentationPredictor.postprocess()`](https://docs.ultralytics.com/reference/models/yolo/semantic/predict).
